server/core/src/main/scala/org.apache.daffodil.debugger.dap/Parse.scala

```diff
@@ -176,71 +180,333 @@ object Parse {
               }
           }
-        ).parMapN(LaunchArgs.apply)
-    }
+      }
+
+    // The following functions granularly parse the tdmlConfig object from the launch config
+    //
+    // tdmlConfig: {
+    //   action: '',
+    //   name: '',
+    //   description: '',
+    //   path: ''
+    // }
+    //
+    // The action field is parsed elsewhere. If these functions are hit, a valid action
+    //   other than 'none' was found.
+
+    // Parse the  name field from the tdmlConfig object from the launch config
+    // Returns an error if the field is missing or is an empty string
+    //
+    // tdmlConfig: tdmlConfig object from the launch config
+    def parseTDMLName(tdmlConfig: JsonObject) =
+      Option(tdmlConfig.getAsJsonPrimitive("name"))
+        .toRight("missing 'tdmlConfig.name' field from launch request")
+        .map(_.getAsString())
+        .flatMap(name => Either.cond(name.length() > 0, name, "'name' field from 'tdmlConfig' object cannot be empty"))
+        .toEitherNel
+
+    // Parse the description field from the tdmlConfig object from the launch config
+    // Returns an error if the field is missing or is an empty string
+    //
+    // tdmlConfig: tdmlConfig object from the launch config
+    def parseTDMLDescription(tdmlConfig: JsonObject) =
+      Option(tdmlConfig.getAsJsonPrimitive("description"))
+        .toRight("missing 'tdmlConfig.description' field from launch request")
+        .map(_.getAsString())
+        .flatMap(description =>
+          Either
+            .cond(description.length() > 0, description, "'description' field from 'tdmlConfig' object cannot be empty")
+        )
+        .toEitherNel
+
+    // Parse the path field from the tdmlConfig object from the launch config
+    // Returns an error if the field is missing or is an invalid path
+    // A case where the user expects a new directory to be created is not a valid path
+    //   eg. /path/to/<existing>/<non-existing>/file.tdml
+    //
+    // tdmlConfig: tdmlConfig object from the launch config
+    def parseTDMLPath(tdmlConfig: JsonObject) =
+      Option(tdmlConfig.getAsJsonPrimitive("path"))
+        .toRight("missing 'tdmlConfig.path' field from launch request")
+        .flatMap(path =>
+          Either
+            .catchNonFatal(Paths.get(path.getAsString).toFile().getAbsolutePath())
+            .leftMap(t => s"'infosetOutput.path' field from launch request is not a valid path: $t")
+            .ensureOr(file => s"can't write to infoset output file at ${file}") { f =>
+              val file = Paths.get(f).toFile()
+              // If an empty string is passed in, it will be set to the workspace directory by default
+              // This is inside the Java code, so we have to make sure that the TDML file we
+              //   are working with is not a directory
+              !file.isDirectory() && (file.canWrite || (!file.exists && file.getParentFile.canWrite))
+            }
+        )
+        .toEitherNel
```

Comment:
I'm a little concerned that this duplicates a decent amount of logic from the daffodil tdml library. And there's a lot of non trivial code in the TDML library that isn't implemented that we definitely don't want to reimplement. For example, input data can be defined in binary, hex, or characters, and even more complicates like bitOrder, replaceDFDLEntitites, etc. And this is just the data portion. There's a lot of complexity to TDML files that I would hate to have duplicated.

Reply:
The intent of this feature was just to implement very basic TDML test cases into the extension. I agree that the code should not be duplicated. I'm hoping that someone will be willing to come along to refactor this to reference the TDML code in the main daffodil repo (either directly, or refactor it into a separate library), but that seemed out of scope for this implementation.

The main use case for this feature as it exists now is for a user of the extension to use the 'Generate TDML' command to create a TDML file that can be added to a DFDL schema and an input file, and sent to the user list for debugging purposes. Then, someone should be able to take those files and execute the test case from the TDML file to get the same execution as the original poster.